akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -422,6 +425,114 @@ final case class Fold[In, Out](zero: Out, f: (Out, In) ⇒ Out) extends GraphSta
 /**
  * INTERNAL API
  */
+final class FoldAsync[In, Out](zero: Out, f: (Out, In) ⇒ Future[Out]) extends GraphStage[FlowShape[In, Out]] {
+  import akka.dispatch.ExecutionContexts
+
+  val in = Inlet[In]("FoldAsync.in")
+  val out = Outlet[Out]("FoldAsync.out")
+  val shape = FlowShape.of(in, out)
+
+  override def toString: String = "FoldAsync"
+
+  override val initialAttributes = DefaultAttributes.foldAsync
+
+  def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler with OutHandler {
+    val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
+
+    private var inFlight = 0
+
+    private var aggregator: Out = zero
+    private var aggregating: Future[Out] = Future.successful(aggregator)
+
+    private def onResume(t: Throwable): Unit = {
+      aggregator = zero
+    }
+
+    private def ec = ExecutionContexts.sameThreadExecutionContext
+
+    private val futureCB = getAsyncCallback[Try[Out]]((result: Try[Out]) ⇒ {
+      inFlight -= 1
+
+      result match {
+        case Success(update) if update != null ⇒ {
+          aggregator = update
+
+          if (isAvailable(out) && !hasBeenPulled(in)) tryPull(in)
+        }
+
+        case other ⇒ {
+          val ex = other match {
+            case Failure(t) ⇒ t
+            case Success(s) if s == null ⇒
+              ReactiveStreamsCompliance.elementMustNotBeNullException
+          }
+
+          if (decider(ex) == Supervision.Stop) failStage(ex)
+          else if (isClosed(in) && inFlight == 0) completeStage()
+          else if (!hasBeenPulled(in)) {
+            onResume(ex)
+            tryPull(in)
+          }
+        }
+      }
+    }).invoke _
+
+    def onPush(): Unit = {
+      try {
+        aggregating = f(aggregator, grab(in))
+        inFlight += 1
+
+        aggregating.value match {
+          case Some(result) ⇒ futureCB(result) // already completed
+          case _            ⇒  aggregating.onComplete(futureCB)(ec)
+        }
+      } catch {
+        case NonFatal(ex) ⇒ decider(ex) match {
+          case Supervision.Stop ⇒ failStage(ex)
+          case _                ⇒ onResume(ex) // Resume or Restart
+        }
+      }
+
+      if (inFlight < 1) tryPull(in)
+    }
+
+    override def onUpstreamFinish(): Unit = {
+      def watchTermination = getAsyncCallback[Try[Out]]((result: Try[Out]) ⇒ {
+        result match {
+          case Failure(ex) ⇒ {
+            if (decider(ex) == Supervision.Stop) failStage(ex)
```

Comment:
So if the other callback "wins" the behavior will be different, it will complete the stage without first emitting the aggregate since:
```
else if (isClosed(in) && inFlight == 0) completeStage()
```
(as it is in the aggregate callback the inFlight check is not needed, that there is no future in flight is implied as far as I can see)